### ISSUE TITLE
Timout issue when uploading large files

### ISSUE THREAD

**Issue by lparsons-42** [elabftw/elabimg#7]
We installed elabftw on an ubuntu (18.04.1) system in our environment recently.  We have projects that have large files (100s of MB to multiple GB) associated with them and we would like to be able to upload these files to their respective projects.  The default file upload restriction appears to be 100 MiB.  

I was advised to change "MAX_UPLOAD_SIZE" to address this.  In /etc/elabftw.yml I found a parameter by this name; first I increased it from "100M" to "10000M", which did not change the restriction.  Someone else suggested I change it to "0" (with no units on it), which I also tried though this also did not make a difference.  I also looked at some yml files in /var/lib/docker but altering them did not change this either, even after rebooting the system.  

I also noticed in the elabftw documentation (https://elabftw.readthedocs.io/en/latest/faq.html) that we were advised to try editing php.ini (which I found in /var/lib/docker/) but this also did not make a difference.  

Either the changes I made to the files are not being considered by elabftw when it starts, or I am editing the wrong files.  I have rebooted the system a couple times along the way in case there was something I needed to do to restart the docker instance to have these read.

thank you

**Comment by NicolasCARPi** [comment]
Sorry the FAQ entry is for a non docker install! I might remove it as clearly it can cause confusion.

Never edit anything from /var/lib/docker !!

I answered you on gitter.im.

![2020-01-21-181207_884x176_scrot](https://user-images.githubusercontent.com/3043706/72826661-8ba86600-3c79-11ea-818a-95116618930b.png)

Rebooting the server is not the same as restarting the containers with `elabctl restart`. (note: you never need to restart the entire server).

**Comment by lparsons-42** [comment]
Thank you for the reply.  It looks like the "elabctl restart" was the part I was not finding before; I expected (likely erroneously) that rebooting the server would restart the docker containers as well.  I just changed the yml file to set the max_upload_size to "10000M" and now I can upload large files.  I had it at "0" before and it was actually cutting me off at 2MiB after "elabctl restart".  

Now I can upload my mass spec files into experiments.

Thank you!

**Comment by NicolasCARPi** [comment]
> I expected that rebooting the server would restart the docker containers

It's a reasonable assumption. What happens when you reboot:

- docker stop containers
- **reboot**
- docker start the same containers again

Whereas with a `elabctl restart`:

- docker stop containers
- docker delete containers
- docker create new containers from the config file

For the "0" case, it's an edge case that I didn't test myself. I think the code is expecting a unit, I'll have a look into it.

**Comment by lparsons-42** [comment]
Thank you that really helps!
I have found over the past couple days that even with the raised limit on file size, large files don't seem to finish uploading to the system.  I'm currently trying to send in files that are ~800MB and they start sending only to stall and never complete.  I'm sending between two systems on the same local WiFi; if I send a smaller (~40MB) file it sends without issue but the larger files even when allowed several hours never progress even to half way in the progress meter.
I'm not sure where to look on this matter.

**Comment by NicolasCARPi** [comment]
What browser are you using? Also, it could be an issue with the WiFi.

**Comment by lparsons-42** [comment]
I have tried several different browsers from within my Windows 10 laptop; Firefox 72.0.2, Chrome 79.0.3945.130, and Edge 44.18362.449.0.  None of them can successfully upload an 800mb file, with the limit turned up to 1000mb.  All browsers will start uploading the file and eventually stall out without finishing or even progressing even if left many hours to do so.  The largest file I can consistently upload is ~30MB.  I have tried a file ~122MB and have had inconsistent results uploading it, without any browser that can do it repeatedly.  
Is there a parameter that sets how much space can be used by the docker container perhaps?  Or anything else that might prevent it from being able to store larger files?
thank you

**Comment by lparsons-42** [comment]
As a follow-up I tried to upload the largest (~811MB) file from the system that is hosting elabftw to itself through the web interface.  In this case the system is running Firefox 72.0.2 (64bit) in Ubuntu 18.04.3.  I found the same behavior here as I have seen when attempting to use another system's browser, in that the file starts to transfer but never finishes.  It does not give an error, it just stalls out without making additional progress.  
Is there anything else I am missing that might be driving this?

**Comment by NicolasCARPi** [comment]
If you open the console with F12, is there an error in there?

If you do `elabctl php-logs`, is there an error related to the upload?

AFAIK, changing the MAX_UPLOAD_SIZE is the only thing you need to do. Do you have a proxy server in front of eLab?

**Comment by lparsons-42** [comment]
Unfortunately to the best I can tell the answer is no to all of your questions.

I just tried uploading the 811MB file this morning, and it has already stalled out.  I hit F12 in firefox while uploading it and have not seen any errors in the console.  

When I ran 'elabctl php-logs' on the system running elabftw it only showed a couple messages from a couple weeks ago, none of which were relevant to this.  

And to the best of my knowledge there is no proxy running on this system.  I'm pretty sure Ubuntu doesn't install one by default, and squid - which to my understanding is the most commonly used proxy in linux - is not installed on this system at all.

**Comment by lparsons-42** [comment]
I think I have ruled out our network as the source of the problem by doing the following:
1.  scp'd the files from my own system to the PC that runs elabftw
2.  logged in to said system and started firefox locally
3.  connected to elabftw in firefox by pointing browser to "https://localhost:443"
4.  logged in and started a new experiment, then attempted to upload a 800MB file
5.  File reaches ~50% fairly quickly, then stalls out indefinitely.  
6.  Starting a console in firefox with F12 gives no relevant console messages - only messages pertaining to the certificate that I have not yet dealt with.

In other words, here a file that is already local to the elabftw server won't finish uploading even when the network is ruled out.  This is 64bit Linux Firefox 72.0.02 in Ubuntu 18.04.3 LTS.

**Comment by NicolasCARPi** [comment]
Unfortunately, I currently have no idea why you encounter this issue :/

Maybe try to have `- MAX_UPLOAD_SIZE=1G` in the conf file and restart the container. Also, can you check that the `php.ini` file has the correct value?

See the command in my screenshot of gitter.im above.

**Comment by NicolasCARPi** [comment]
maybe it's important to have only 'm' not 'mb'.

**Comment by lparsons-42** [comment]
I was wondering if maybe it didn't like that the value ending in "M" was in the "G" range so I changed it to "100G" and reset the system with "elabctl restart".  
Then I ran
docker exec -it elabftw grep "upload_max_filesize" /etc/php7/php.ini
and it returned
upload_max_filesize = 100G
Then I tried to upload the 800MB file and it stalled out again, even from the local system that connected via firefox at https://localhost:431.  Console (F12) does not show any useful feedback.  Neither does "elabctl php-logs" or /var/log/syslog.

**Comment by NicolasCARPi** [comment]
So the good news is that I am able to reproduce the issue with a distant server. Meaning that it works LOCALLY on my dev box, but on a server, it doesn't.

I even tried setting `client_max_body_size` to 0 in nginx.conf but it's still stalling. As you mentioned, it is not a browser issue, nor a PHP one it seems.

If the request was too large, nginx would reply with a 413 code. But here it's stalling, and there are NO errors logged anywhere....

So the most likely culprit is now the Dropzone.js library dealing with the upload.

It turns out that there is a `timeout` setting that I was unaware of. By default it is 30s and it is exactly the time after which it stalls. See: https://stackoverflow.com/questions/44512084/large-file-upload-request-gets-cancelled

Because my test server only has 1G of RAM I am not able to rebuild the JS, so I launched a build on the hypernext branch with docker hub, it'll take like 15 min to build but I wanted to post this answer before confirming that this is indeed the issue at play!

I'll give you the steps to follow to increase this timeout in my next message.

**Comment by NicolasCARPi** [comment]
Ok so docker hub is having network issues..... :/

Steps to try it yourself:

1. `docker exec -it elabftw bash`
2. `apk add --update vim` Alternatively, you can install nano and use that to edit the file
3. `vim web/app/js/src/edit.js`
4. Use arrow keys to go down to [line 15](https://github.com/elabftw/elabftw/blob/master/web/app/js/src/edit.js#L15)
5. Press `o` to insert a new line and add `timeout: 300000,`
5b. Press ESC to get out of INSERT mode
6. `:x` to exit vim
7. `yarn install`
8. `yarn buildall`

That should do it.

**Comment by NicolasCARPi** [comment]
Ok it looks like this WAS the issue at play! The fix is already on the hypernext branch and will be part of the next release. In the mean time, see my message above ;)

**Comment by lparsons-42** [comment]
Interesting!  I'm glad I was able to help you find this.  Unfortunately it hasn't helped my case.  I followed your "docker exec" commands and then restarted elabftw but my uploads still stall out indefinitely after ~30 seconds.  I even added an extra 0 to "timeout:" but that made no difference.  I'll watch for the container update, hopefully that will resolve it in ways that I just can't see here.

**Comment by NicolasCARPi** [comment]
That's because you must not restart the containers! ;)

When you restart the container, all the changes you made are gone. Here you are editing the live running container (something that should never be done).

**Comment by lparsons-42** [comment]
I guess I spend too much time using Windows; I'm used to restarting as the solution.  Indeed with the server left un-restarted after the edits, the files do now post.  Thank you for your work on this!

**Comment by lparsons-42** [comment]
Is the new release available now?  I don't know the release schedule.  If so how do I update?  I'm still very new to the Docker way of doing things.  
thank you!
Lee

**Comment by NicolasCARPi** [comment]
No it's not out. Use github's watch mechanism to watch for realeases. An alpha version should be out soon.

**Comment by NicolasCARPi** [comment]
Alpha version is out with this fix.
